packages/hash/frontend/src/middleware/return-types-as-json.ts

```diff
@@ -0,0 +1,139 @@
+import {
+  DataType,
+  EntityType,
+  PropertyType,
+  TypeSystemInitializer,
+  validateVersionedUri,
+} from "@blockprotocol/type-system";
+import { apiGraphQLEndpoint } from "@hashintel/hash-shared/environment";
+import { NextResponse } from "next/server";
+
+import type { NextRequest } from "next/server";
+
+import type { ApolloError } from "apollo-server-express";
+import type {
+  GetDataTypeQuery,
+  GetDataTypeQueryVariables,
+  GetEntityTypeQuery,
+  GetEntityTypeQueryVariables,
+  GetPropertyTypeQuery,
+  GetPropertyTypeQueryVariables,
+} from "../graphql/apiTypes.gen";
+import { generateQueryArgs } from "./return-types-as-json/generate-query-args";
+
+const generateErrorResponse = (
+  status: 400 | 401 | 404 | 500,
+  message: string,
+) =>
+  new NextResponse(
+    JSON.stringify({
+      error: message,
+    }),
+    { status, headers: { "content-type": "application/json" } },
+  );
+
+const generateJsonResponse = (object: DataType | EntityType | PropertyType) =>
+  new NextResponse(JSON.stringify(object, undefined, 2), {
+    headers: { "content-type": "application/json" },
+  });
+
+const makeGraphQlRequest = async <Data, Variables>(
+  query: string,
+  variables: Variables,
+  cookie: string | null,
+): Promise<{ data?: Data | null; errors?: ApolloError[] | null }> => {
+  const { data, errors } = await fetch(apiGraphQLEndpoint, {
+    method: "POST",
+    headers: { "content-type": "application/json", cookie: cookie ?? "" },
+    body: JSON.stringify({ query, variables }),
+  }).then((resp) => resp.json());
+
+  return { data, errors };
+};
+
+export const returnTypeAsJson = async (request: NextRequest) => {
+  // @todo this blows up the middleware function's compiled size – is that a problem? is it worth it?
+  await TypeSystemInitializer.initialize();
+
+  const { url } = request;
+
+  const { inner: validationResult, type: validationResultType } =
+    validateVersionedUri(url);
+
+  if (validationResultType === "Err") {
+    return generateErrorResponse(
+      400,
+      `Requests for JSON representations of types must be made to a versioned URI, e.g. ${url.replace(
+        /\/$/,
+        "",
+      )}/v/1`,
+    );
+  }
```

Comment:
Just flagging that there are a lot of error variants for the versioned URI:
https://github.com/blockprotocol/blockprotocol/blob/main/crates/type-system/src/ontology/uri/error.rs#L21-L36
do we care about making the response more precise, or is close enough good enough?

Reply:
If we don't implement these now we should capture them in a follow-up.

Reply:
I'd expect the error to be `InvalidBaseUri` in that case – you've explained that http://localhost:3000/@example-org/types/entity-type/user is invalid because it's missing a trailing slash

I don't think we need to sink time into this now though, "it's wrong" is fine